### ISSUE TITLE
[PAM Password quality] change password allowed using wrong credentials 

### ISSUE THREAD

**Issue by gangaram-tii** [NixOS/nixpkgs#287420]
### Describe the bug
I am using following rules to enforce strong password. Though it enforces strong password but it allows me to change the password using wrong current password.

```
security.pam.services.passwd.rules.password.pwquality = {
        control = "required"; 
        modulePath = "${pkgs.libpwquality.lib}/lib/security/pam_pwquality.so"; 
        # order BEFORE pam_unix.so
        order =  config.security.pam.services.passwd.rules.password.unix.order - 10;
        settings = {
          retry = 3;
          minlen = 8;
          difok = 6;
          dcredit = -1;
          ucredit = 1;
          ocredit = -1;
          lcredit = 1;
          enforce_for_root = true;
        }; 
      };

```

### Steps To Reproduce
1. Use above configuration
2. Run passwd command
3. Enter wrong current password
4. Enter new password
Password changed successfully.

### Expected behavior
I should not be allowed to change password using wrong current password.

### Metadata

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.1.47, NixOS, 23.05 (Stoat), 23.05.3085.2ab91c8d65c0`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.13.5`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

**Comment by aikooo7** [comment]
I'm not really understanding what you mean by "wrong current password", that means you can just change the password using any password? So it is not checking if the password is correct?

**Comment by gangaram-tii** [comment]
Yes

**Comment by felbinger** [comment]
I could confirm this behavior on NixOS 24.05

**Comment by NotAShelf** [comment]
I cannot confirm this behaviour on NixOS 24.11. Upon entering a wrong password, followed by a new password and the confirmation of the new password you will receive a token manipulation error.

**Comment by felbinger** [comment]
That's weird, because the issue still exist on my system:
```sh
sudo adduser test
echo 'test:C51.kH9wGnwHwJVRkN19' | sudo chpasswd
sudo su test
passwd
```
```
Changing password for test.
Current password: NotTheRealPassword
New password: AnyThingThatFits-P0licy
Retype new password: AnyThingThatFits-P0licy
passwd: password updated successfully
```
```
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.8.12, NixOS, 24.11 (Vicuna), 24.11.20240616.b60ebf5`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.22.1`
 - nixpkgs: `/nix/store/5w7nhyvrgsa81h0kyifmbl6j97hvlzh5-source`
```

**Comment by NotAShelf** [comment]
You're right, I had another PAM rule that caused an override. Removing it would cause `passwd` to change passwords without the old password.

I think the line responsible is

```
password sufficient pam_unix.so nullok yescrypt # unix (order 10200)
```

in `/etc/pam.d/passwd`. Other distros have this set to `required`, but NixOS seems to be defaulting to `sufficient` since 19.05.

**Comment by felbinger** [comment]
@Majiir as the maintainer of the pam module, how should we proceed? Is this an actual error in the module that needs to be investigated, or have we just made a mistake in the configuration from above?

**Comment by felbinger** [comment]
@NotAShelf But without the `pwquality` pam module, the line you mentioned is also included in the `/etc/pam.d/passwd`. Without loading the above configuration the old password is still required, so I'm wondering why you think that this line causes the issue :thinking:

**Comment by NotAShelf** [comment]
My guess was kinda random, making comparisons to my Arch box and some man pages of PAM. In my testing, I got to override that `sufficient` with *`required`*  - which lead to a different error.
